feat: verification email

rayrunzewang opened the pull request:
Added email verification function
- send verification email after user registration
- a verification email should be sent to the user's provided email address (the email contains a link/button(valid only 30 mins) for email verification)
- verify through email link
- when a new user click on the verification link (”Verify now”) Then a user's account should be verified
- if Link expire, will send another verification email
- if a unverified user attempt to login, will require verification and send verification email

Resolve CP-8

![2024-05-16 11_17_52-README md - platform_api - Visual Studio Code](https://github.com/BeeQuantAI/platform_api/assets/80236130/c2a41089-afe8-4577-a807-4e5ad44a043b)
![2024-05-23 04_43_42-Verify Your Email - rayrunzewang@gmail com - Gmail](https://github.com/BeeQuantAI/platform_api/assets/80236130/dd9b47fb-ffa1-4f77-8f98-c7a52242fad2)
![2024-06-15 11_30_45-Greenshot](https://github.com/BeeQuantAI/platform_api/assets/80236130/5786f853-75a3-4cf9-a3c4-cb7ccb431e58)
![2024-06-15 11_24_37-Greenshot](https://github.com/BeeQuantAI/platform_api/assets/80236130/a70388ed-41f9-48d0-9541-ee2d93c9853e)
![2024-06-15 11_28_22-Greenshot](https://github.com/BeeQuantAI/platform_api/assets/80236130/2a6a2926-b67c-45e7-a0ce-69e08102cd92)

Chris-Wang commented on .github/workflows/CI.yml:
Need to revert changes

Chris-Wang commented on .github/workflows/Uat-2-Prod-Boris.yml:
Need to revert changes

Chris-Wang commented on .github/workflows/Uat-Deploy-Boris.yml:
Need to revert changes

Chris-Wang commented on .github/workflows/images/readmeScreenshot-uatDeployWorkflow.png:
Need to revert changes

Chris-Wang commented on .github/workflows/images/readmeScreenshot-uatDeployterraformWorkflow.png:
Need to revert changes

Chris-Wang commented on README.md:
Need to revert changes

Chris-Wang commented on src/modules/auth/auth.emailVerificationService.ts:
```diff
@@ -0,0 +1,56 @@
+  async sendVerificationEmail(user: User): Promise<Result> {
+    const transporter = nodemailer.createTransport({
+      service: 'Gmail',
+      host: 'smtp.gmail.com',
```
Our domain used Microsoft 365, needs to update service and host

Chris-Wang commented on src/modules/auth/auth.emailVerificationService.ts:
```diff
@@ -0,0 +1,56 @@
+    try {
+
+      await transporter.sendMail({
+        from: 'baysidetechstudio@gmail.com',
```
from email needs to be updated to `info@beequant.ai`

Chris-Wang commented on src/modules/auth/auth.emailVerificationService.ts:
needs to rename this file as `email.service.ts` considering it will be used in other services

Chris-Wang commented on src/modules/auth/auth.emailVerificationService.ts:
```diff
@@ -0,0 +1,56 @@
+      },
+    });
+
+    const verificationLink = `http://localhost:5173/verify-email?email=${user.email}&token=${user.verificationToken}`;
```
for this temp url, needs to add `TODO: link will be updated once api server is deployed, currently this url is for local development only`

Chris-Wang commented on src/modules/auth/auth.emailVerificationService.ts:
```diff
@@ -0,0 +1,56 @@
+        to: user.email,
+        subject: 'Verify Your Email',
+        html: `
+        <div style="font-family: Arial, sans-serif; padding: 20px; color: #007bff">
```
probably attach a screenshot of the email

Chris-Wang commented on src/modules/auth/auth.service.ts:
```diff
@@ -53,20 +76,127 @@ export class AuthService {
+
+  async verifyEmail(email: string, token: string): Promise<Result> {
+    try {
+      const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
```
regex needs to be imported and keep the same through all different services for email pattern

Chris-Wang commented on src/modules/auth/auth.service.ts:
```diff
@@ -53,20 +76,127 @@ export class AuthService {
+      }
+
+      if (user.verificationToken === null) {
+
```
unused row

Chris-Wang reviewed (changes requested):
Needs to resolve conflict

Maaayhan committed:
feat: verification email
Added email verification function

- send verification email after user registration
- a verification email should be sent to the user's provided email address (the email contains a link/button(valid only 30 mins) for email verification)
- verify through email link
- when a new user click on the verification link (”Verify now”) Then a user's account should be verified
- if Link expire, will send another verification email
- if a unverified user attempt to login, will require verification and send verification email

Resolve CP-8

Chris-Wang reviewed (approved):
LGTM
